src/node/interfaces.cpp

```diff
@@ -938,9 +938,71 @@ class BlockTemplateImpl : public BlockTemplate
         return chainman().ProcessNewBlock(block_ptr, /*force_processing=*/true, /*min_pow_checked=*/true, /*new_block=*/nullptr);
     }
 
+    std::unique_ptr<BlockTemplate> waitNext(CAmount fee_threshold, MillisecondsDouble timeout) override
+    {
+        if (timeout > std::chrono::years{100}) timeout = std::chrono::years{100}; // Upper bound to avoid UB in std::chrono
+
+        CAmount current_fees = 0;
+        if (fee_threshold < MAX_MONEY) {
+            for (CAmount fee : m_block_template->vTxFees) {
+                // Skip coinbase
+                if (fee < 0) continue;
+                current_fees += fee;
+            }
+        }
+
+        // Alternate waiting for a new tip and checking if fees have risen.
+        // The latter check is expensive so we only run it once per second.
+        auto now{std::chrono::steady_clock::now()};
+        const auto deadline = now + timeout;
+        const MillisecondsDouble tick{1000};
+        std::unique_ptr<BlockTemplateImpl> block_template;
+
+        while (now < deadline) {
+            {
+                WAIT_LOCK(notifications().m_tip_block_mutex, lock);
+                notifications().m_tip_block_cv.wait_until(lock, std::min(now + tick, deadline), [&]() EXCLUSIVE_LOCKS_REQUIRED(notifications().m_tip_block_mutex) {
+                    return notifications().m_tip_block != m_block_template->block.hashPrevBlock || chainman().m_interrupt;
+                });
+            }
+
+            if (chainman().m_interrupt) return nullptr;
+
+            // Must release m_tip_block_mutex before locking cs_main, to avoid deadlocks.
+            LOCK(::cs_main);
+
+            // The only way to determine if fees increased compared to the previous template,
+            // is to generate a fresh template. Cluster Mempool may allow for a more efficient
+            // way to determine how much (approximate) fees for the next block increased.
+            block_template = std::make_unique<BlockTemplateImpl>(m_script_pub_key, m_assemble_options, BlockAssembler{chainman().ActiveChainstate(), context()->mempool.get(), m_assemble_options}.CreateNewBlock(m_script_pub_key), m_node);
+
+            // If the tip changed, return the new template regardless of its fees.
+            if (block_template->m_block_template->block.hashPrevBlock != m_block_template->block.hashPrevBlock) {
+                return block_template;
+            }
+
+            CAmount new_fees = 0;
+            for (CAmount fee : block_template->m_block_template->vTxFees) {
+                // Skip coinbase
+                if (fee < 0) continue;
+                new_fees += fee;
+                if (new_fees >= current_fees + fee_threshold) return block_template;
+            }
+            //
+            block_template.reset();
+        }
+
+        return block_template;
+    }
+
+    const CScript m_script_pub_key;
```

Comment:
Separate from this PR, I'm looking into whether we can avoid passing `script_pub_key` around. Afaik it's only used by tests and for solo CPU mining.